eli5/keras/gradcam.py

```diff
@@ -90,58 +36,66 @@ def gradcam_backend(model, # type: Model
     targets : list, optional
         Index into the network's output,
         indicating the output node that will be
-        used as the "loss" during differentiation.
+        used as the "loss" or scalar target during differentiation.
 
     activation_layer : keras.layers.Layer
         Keras layer instance to differentiate with respect to.
-    
 
-    See :func:`eli5.keras.explain_prediction` for description of the 
-    ``model``, ``doc``, ``targets`` parameters.
+
+    See :func:`eli5.keras.explain_prediction` for description of the
+    ``model``, ``doc``, and ``targets`` parameters.
 
     Returns
     -------
-    (weights, activations, gradients, predicted_idx, predicted_val) : (numpy.ndarray, ..., int, float)
+    (activations, gradients, predicted_idx, predicted_val) :
+        (numpy.ndarray, numpy.ndarray, int, float)
         Values of variables.
     """
     # score for class in targets
-    predicted_idx = _get_target_prediction(targets, model)
-    predicted_val = K.gather(model.output[0,:], predicted_idx) # access value by index
+    # TODO: maybe do the sum / loss calculation in this function and pass it to gradcam.
+    # This would be consistent with what is done in
+    # https://github.com/ramprs/grad-cam/blob/master/misc/utils.lua
+    # and https://github.com/ramprs/grad-cam/blob/master/classification.lua
+    # TODO: as in pytorch PR, separate out classification tensor code
+    if targets is not None:
+        target, = targets
+        predicted_idx = K.constant([target], dtype='int64')
+    else:
+        predicted_idx = _autoget_target_prediction(model)
+    # access value by index
+    predicted_val = K.gather(model.output[0, :], predicted_idx)
 
-    # output of target activation layer, i.e. activation maps of a convolutional layer
-    activation_output = activation_layer.output 
+    # output of target activation layer, i.e. activation maps of a conv layer
+    activation_output = activation_layer.output
 
     # score for class w.r.p.t. activation layer
     grads = _calc_gradient(predicted_val, [activation_output])
 
-    # Global Average Pooling of gradients to get the weights
-    # note that axes are in range [-rank(x), rank(x)) (we start from 1, not 0)
-    # TODO: decide whether this should go in gradcam_backend() or gradcam()
-    weights = K.mean(grads, axis=(1, 2))
+    # TODO: gradcam on input layer
```

Comment:
Could you please explain this TODO item in more detail, does this refer to some functionality that is not implemented yet, or something else?

Reply:
If I try to do an explanation using the first layer from mobilenet (model in the image tutorial), I get the following:
![Screenshot from 2019-08-19 18-33-52](https://user-images.githubusercontent.com/39887323/63287085-5b673080-c2b1-11e9-85fa-62e1f2660d57.png)

Essentially this error happens when I am using `K.function` and pass an input that is also an output ("both fed and fetched").

Regarding use cases: maybe generating adversarial images is one use?